lib/library.cpp

```diff
@@ -1257,7 +1257,7 @@ bool Library::isCompliantValidationExpression(const char* p)
         if (std::isdigit(*p))
             error |= (*(p + 1) == '-');
         else if (*p == ':') {
-            error |= range | (*(p + 1) == '.');
+            error |= range || (*(p + 1) == '.');
```

Comment:
this changes the sematics do we really want to have an extra condition here.. I am not sure we should activate the bitwiseOnBoolean in Cppcheck itself. I believe the main motivation is stylistic when people dislike to treat booleans as integers. And personally I don't have that problem with this.

Reply:
Not sure what the change in semantics is in this case. Generally, binary `|` does not short-circuit, so if this is allowed, code will probably creep in which relies on side effects of the second condition.

Reply:
maybe we should move bitwiseOnBoolean to an addon.

Reply:
Isn't that just what severity `style` is intended for (which bitwiseOnBoolean has)? Addons don't seem very popular except the MISRA one...

Reply:
the motivation for `style` checkers in core cppcheck should be to find bugs and I would say it should have a relatively high acceptance in the user community.

I am afraid that `|` is often used by intention often. As far as I know it is not widely considered bad practice / dangerous to use it.

Reply:
Undo this change?

Reply:
It seems like the check should be a performance check rather then a style check. The reason `||` is preferred is because it will short-circuit(and avoid extra unnecessary calculations). The only reason to use `|` with booleans is if it there is side effects and thus short-circuiting will not be equivalent. 

I think for cppcheck's bitwiseOnBoolean it should check for side effects using `isConstExpression`, such checks like that cannot be done in an addon, so it needs to be done in cppcheck. 

A more general check(ignoring side-effects) could be added to the cert.py addon as cert doens't have such an exception and in order to show it is intentional it requires extra set of parens.

Reply:
> Undo this change?

Yes please

> The reason || is preferred is because it will short-circuit(and avoid extra unnecessary calculations)

I believe that the original motivation was that some people doesn't like to use bitwise on booleans for stylistic reasons. At that time we did not have addons. If that is the motivation then I feel that this should be moved to an addon however I guess I don't want to do that right now. Let's wait until after the release.

If we want to warn about a performance issue then I think the check should try to determine if the lhs condition is likely true. and if the rhs might be heavy. A purely semantic check of types is too simple.

I am no expert but I would assume that blindly replacing all `|` on boolean expressions with `||` can sometimes lead to slower code.

Reply:
> I believe that the original motivation was that some people doesn't like to use bitwise on booleans for stylistic reasons.

It does make the code much clearer as you can see the operations on booleans and have no side effects, whereas using `|` looks like you are not using boolean and the expressions may have side effects. This is the same reason why cppcheck tells users to make variables `const` with the constVariable check. Leaving the variable as non-const is still correct, just as using `|` is still correct, but the difference is that users can more clearly see the code doesn't have side effects.

And this extra clarity in code can also be used to find bugs. If the intention of the user is to have side effects and then cppcheck complains about constVariable or bitwiseOnBoolean then obviously something is wrong(and the solution in this case is not to add `const` or `||` but instead ensure the side effects are being properly called).

> If that is the motivation then I feel that this should be moved to an addon

If we want a check that avoids FPs then it cannot be moved to an addon. A addon cannot call `isConstExpression`. 

And sure the current check doesn't call `isConstExpression`, but it should to avoid FPs. So if we move this to an addon then we wont be able to fix those FPs. So this should definitely NOT be moved to an addon.

Reply:
> It does make the code much clearer as you can see the operations on booleans and have no side effects

I believe the original motivation was to always warn when booleans was used no matter if the expressions have side effects or not. It was by design to warn about something like `b1|b2` where you just read two ordinary variables. I guess if you come from Java or something then that feels better. That rule can easily be moved to an addon.

If we can enhance this check so its focus will be to detect unsafe code; that is more interesting. Of course, that check does not have to be moved to an addon. However that would mean false negatives for those that still want to have the semantic warnings.

So I think we can easily implement the bitwiseOnBoolean in an addon as-is.. and it sounds ok to add some proper heuristics in the c++ code and rename the c++ check to unsafeBitwiseOnBoolean or something like that.